finalitygadget/finalitygadget.go

```diff
@@ -499,43 +574,116 @@ func (fg *FinalityGadget) queryBlockByHeight(blockNumber int64) (*types.Block, e
 	}, nil
 }
 
-func (fg *FinalityGadget) handleBlock(ctx context.Context, latestFinalizedHeight uint64) error {
-	for height := fg.lastProcessedHeight + 1; height <= latestFinalizedHeight; height++ {
+type blockResult struct {
+	height      uint64
+	block       *types.Block
+	isFinalized bool
+	err         error
+}
+
+func (fg *FinalityGadget) processBlocksTillHeight(ctx context.Context, latestHeight uint64) error {
+	for startHeight := fg.lastProcessedHeight + 1; startHeight <= latestHeight; {
 		select {
 		case <-ctx.Done():
 			return nil
 		default:
-			if height > math.MaxInt64 {
-				return fmt.Errorf("block height %d exceeds maximum int64 value", height)
+			// Calculate batch start and end heights
+			endHeight := startHeight + fg.batchSize - 1
+			if endHeight > latestHeight {
+				endHeight = latestHeight
 			}
-			block, err := fg.queryBlockByHeight(int64(height))
-			if err != nil {
-				return fmt.Errorf("error getting block at height %d: %w", height, err)
+			fg.logger.Info("Processing batch of blocks", zap.Uint64("start_height", startHeight), zap.Uint64("end_height", endHeight))
+
+			// Create batch of blocks to check in parallel
+			results := make(chan blockResult, endHeight-startHeight+1)
+			var wg sync.WaitGroup
+
+			// Query batch in parallel
+			for height := startHeight; height <= endHeight; height++ {
+				if height > math.MaxInt64 {
+					return fmt.Errorf("block height %d exceeds maximum int64 value", height)
+				}
+				wg.Add(1)
+				go func(h uint64) {
+					defer wg.Done()
+					results <- fg.processHeight(h)
+				}(height)
 			}
 
-			// Check the block is babylon finalized using sdk client
-			isFinal, err := fg.QueryIsBlockBabylonFinalized(block)
-			if err != nil && !errors.Is(err, types.ErrBtcStakingNotActivated) {
-				return fmt.Errorf("error checking block %d: %v", block.BlockHeight, err)
+			// Close results channel once all goroutines complete
+			go func() {
+				wg.Wait()
+				close(results)
+			}()
+
+			// Process results and extract error (if any)
+			statusMap := make(map[uint64]blockResult)
+			var processingError error
+			for result := range results {
+				if result.err != nil {
+					processingError = result.err
+					break
+				}
+				statusMap[result.height] = result
 			}
-			// If not finalized, throw error
-			if !isFinal {
-				return fmt.Errorf("block %d should be finalized according to client but is not", block.BlockHeight)
+			if processingError != nil {
+				return processingError
```

Comment:
in the goroutine processing heights above, we could add a context with cancel to cancel all goroutines as long as there is one goroutine finished with error

but it's probably a low-pri. we can just add a TODO comment